### ISSUE TITLE
Linker errors with Maya 2016 and VS 2015

### ISSUE THREAD

**Issue by koenvandensteen** [chadmv/cvwrap#8]
Hi I cannot figure out why I got the following linker errors: 

`2>cvWrapDeformer.obj : error LNK2001: unresolved external symbol "public: static struct _cl_command_queue * __cdecl MOpenCLInfo::getMayaDefaultOpenCLCommandQueue(void)" (?getMayaDefaultOpenCLCommandQueue@MOpenCLInfo@@SAPEAU_cl_command_queue@@XZ)
2>cvWrapDeformer.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class MVector __cdecl MVector::operator*(float)const " (__imp_??DMVector@@QEBA?AV0@M@Z) referenced in function "public: static long __cdecl CVWrap::EvaluateWrap(void *)" (?EvaluateWrap@CVWrap@@SAJPEAX@Z)
2>cvWrapDeformer.obj : error LNK2001: unresolved external symbol "public: virtual class MStatus __cdecl MPxNode::dependsOn(class MPlug const &,class MPlug const &,bool &)const " (?dependsOn@MPxNode@@UEBA?AVMStatus@@AEBVMPlug@@0AEA_N@Z)`

I have setup my project with the CMake files you provided.

**Comment by redbean** [comment]
I have same issue as above, could you Please comment us for it?
Thank you

**Comment by chadmv** [comment]
I can't seem to reproduce.  I tried with Maya 2016 and Visual Studio Community 2015 Update 3 and it compiled fine.

```
D:\dev\maya\cvwrap\build2016>cmake -G "Visual Studio 14 2015 Win64" -DMAYA_VERSION=2016 ../
-- The C compiler identification is MSVC 19.0.24215.1
-- The CXX compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/x86_amd64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Maya: C:/Program Files/Autodesk/Maya2016/include
-- Configuring done
-- Generating done
-- Build files have been written to: D:/dev/maya/cvwrap/build2016

D:\dev\maya\cvwrap\build2016>cmake --build . --target install --config Release
Microsoft (R) Build Engine version 14.0.25420.1
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 6/17/2017 2:53:05 PM.
Project "D:\dev\maya\cvwrap\build2016\install.vcxproj" on node 1 (default targets).
Project "D:\dev\maya\cvwrap\build2016\install.vcxproj" (1) is building "D:\dev\maya\cvwrap\build2016\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "x64\Release\ZERO_CHECK\".
  Creating directory "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\".
InitializeBuildStatus:
  Creating "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Checking Build System
  CMake does not need to re-run because D:/dev/maya/cvwrap/build2016/CMakeFiles/generate.stamp is up-to-date.
  CMake does not need to re-run because D:/dev/maya/cvwrap/build2016/src/CMakeFiles/generate.stamp is up-to-date.
FinalizeBuildStatus:
  Deleting file "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "D:\dev\maya\cvwrap\build2016\ZERO_CHECK.vcxproj" (default targets).

Project "D:\dev\maya\cvwrap\build2016\install.vcxproj" (1) is building "D:\dev\maya\cvwrap\build2016\ALL_BUILD.vcxproj" (3) on node 1 (default targets).
Project "D:\dev\maya\cvwrap\build2016\ALL_BUILD.vcxproj" (3) is building "D:\dev\maya\cvwrap\build2016\src\cvwrap.vcxproj" (4) on node 1 (default targets).
PrepareForBuild:
  Creating directory "cvwrap.dir\Release\".
  Creating directory "D:\dev\maya\cvwrap\build2016\src\Release\".
  Creating directory "cvwrap.dir\Release\cvwrap.tlog\".
InitializeBuildStatus:
  Creating "cvwrap.dir\Release\cvwrap.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule D:/dev/maya/cvwrap/src/CMakeLists.txt
  CMake does not need to re-run because D:/dev/maya/cvwrap/build2016/src/CMakeFiles/generate.stamp is up-to-date.
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\CL.exe /c /ID:\dev\maya\cvwrap\src\Maya::Maya /I"C:\Program Files\Autodesk\Maya2016\include" /nologo /W3
   /WX- /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D REQUIRE_IOSTREAM /D _BOOL /D NT_PLUGIN /D "CMAKE_INTDIR=\"Release\"" /D cvwrap_EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /MD /G
  S /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"cvwrap.dir\Release\\" /Fd"cvwrap.dir\Release\vc140.pdb" /Gd /TP /errorReport:queue   " /arch:AVX" D:\dev\maya\cvwra
  p\src\pluginMain.cpp D:\dev\maya\cvwrap\src\cvWrapCmd.cpp D:\dev\maya\cvwrap\src\cvWrapDeformer.cpp D:\dev\maya\cvwrap\src\bindingio.cpp D:\dev\maya\cvwrap\src\common.cpp
  pluginMain.cpp
  cvWrapCmd.cpp
  cvWrapDeformer.cpp
  bindingio.cpp
  common.cpp
  Generating Code...
Link:
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\link.exe /ERRORREPORT:QUEUE /OUT:"D:\dev\maya\cvwrap\build2016\src\Release\cvwrap.mll" /INCREMENTAL:NO /
  NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib "C:\Program Files\Autodesk\Maya2016\lib\OpenMaya.
  lib" "C:\Program Files\Autodesk\Maya2016\lib\OpenMayaAnim.lib" "C:\Program Files\Autodesk\Maya2016\lib\OpenMayaFX.lib" "C:\Program Files\Autodesk\Maya2016\lib\OpenMayaRender
  .lib" "C:\Program Files\Autodesk\Maya2016\lib\OpenMayaUI.lib" "C:\Program Files\Autodesk\Maya2016\lib\Foundation.lib" "C:\Program Files\Autodesk\Maya2016\lib\clew.lib" /MANI
  FEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:"D:/dev/maya/cvwrap/build2016/src/Release/cvwrap.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /N
  XCOMPAT /IMPLIB:"D:/dev/maya/cvwrap/build2016/src/Release/cvwrap.lib" /MACHINE:X64   /machine:x64 /export:initializePlugin /export:uninitializePlugin /DLL cvwrap.dir\Release
  \pluginMain.obj
  cvwrap.dir\Release\cvWrapCmd.obj
  cvwrap.dir\Release\cvWrapDeformer.obj
  cvwrap.dir\Release\bindingio.obj
  cvwrap.dir\Release\common.obj
     Creating library D:/dev/maya/cvwrap/build2016/src/Release/cvwrap.lib and object D:/dev/maya/cvwrap/build2016/src/Release/cvwrap.exp
  cvwrap.vcxproj -> D:\dev\maya\cvwrap\build2016\src\Release\cvwrap.mll
FinalizeBuildStatus:
  Deleting file "cvwrap.dir\Release\cvwrap.tlog\unsuccessfulbuild".
  Touching "cvwrap.dir\Release\cvwrap.tlog\cvwrap.lastbuildstate".
Done Building Project "D:\dev\maya\cvwrap\build2016\src\cvwrap.vcxproj" (default targets).

PrepareForBuild:
  Creating directory "x64\Release\ALL_BUILD\".
  Creating directory "x64\Release\ALL_BUILD\ALL_BUILD.tlog\".
InitializeBuildStatus:
  Creating "x64\Release\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule D:/dev/maya/cvwrap/CMakeLists.txt
  CMake does not need to re-run because D:/dev/maya/cvwrap/build2016/CMakeFiles/generate.stamp is up-to-date.
FinalizeBuildStatus:
  Deleting file "x64\Release\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild".
  Touching "x64\Release\ALL_BUILD\ALL_BUILD.tlog\ALL_BUILD.lastbuildstate".
Done Building Project "D:\dev\maya\cvwrap\build2016\ALL_BUILD.vcxproj" (default targets).

PrepareForBuild:
  Creating directory "x64\Release\INSTALL\".
  Creating directory "x64\Release\INSTALL\INSTALL.tlog\".
InitializeBuildStatus:
  Creating "x64\Release\INSTALL\INSTALL.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
PostBuildEvent:
  setlocal
  "C:\Program Files\CMake\bin\cmake.exe" -DBUILD_TYPE=Release -P cmake_install.cmake
  if %errorlevel% neq 0 goto :cmEnd
  :cmEnd
  endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
  :cmErrorLevel
  exit /b %1
  :cmDone
  if %errorlevel% neq 0 goto :VCEnd
  :VCEnd
  -- Install configuration: "Release"
  -- Installing: D:/dev/maya/cvwrap/build2016/./scripts
  -- Installing: D:/dev/maya/cvwrap/build2016/./scripts/AEcvWrapTemplate.mel
  -- Installing: D:/dev/maya/cvwrap/build2016/./scripts/cvwrap
  -- Installing: D:/dev/maya/cvwrap/build2016/./scripts/cvwrap/bindui.py
  -- Installing: D:/dev/maya/cvwrap/build2016/./scripts/cvwrap/menu.py
  -- Installing: D:/dev/maya/cvwrap/build2016/./scripts/cvwrap/__init__.py
  -- Installing: D:/dev/maya/cvwrap/build2016/plug-ins/cvwrap.mll
  -- Installing: D:/dev/maya/cvwrap/build2016/plug-ins/cvwrap.cl
FinalizeBuildStatus:
  Deleting file "x64\Release\INSTALL\INSTALL.tlog\unsuccessfulbuild".
  Touching "x64\Release\INSTALL\INSTALL.tlog\INSTALL.lastbuildstate".
Done Building Project "D:\dev\maya\cvwrap\build2016\install.vcxproj" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:04.93
```

**Comment by redbean** [comment]
Thank you for reply Chad.

Still have same error with your command in cmake as above. 

if you do not mind, I would like to share my environment and how things goes in my work space. 

1. I am using 
visual studio community 2015 update 3
Autodesk® Maya 2016 Ext 2 - Developer Kit (devkit)

2. also using your cgmake and repo for building.
3. cmake commands are 
cmake -G "Visual Studio 14 2015 Win64" -DMAYA_VERSION=2018 ../
cmake --build . --config Release

It would help you to understand.
I will try to build again with ext 1 and tell you how it goes.

Thank you for your pay attention.

**Comment by chadmv** [comment]
You have Maya 2016 ext 2, which I think is 2016.5.  But you are specifying
2018 in the cmake command.

On Jun 18, 2017 7:54 PM, "Jeewon Park" <notifications@github.com> wrote:

Thank you for reply Chad.

Still have same error with your command in cmake as above.

if you do not mind, I would like to share my environment and how things
goes in my work space.

   1.

   I am using
   visual studio community 2015 update 3
   Autodesk® Maya 2016 Ext 2 - Developer Kit (devkit)
   2.

   also using your cgmake and repo for building.
   3.

   cmake commands are
   cmake -G "Visual Studio 14 2015 Win64" -DMAYA_VERSION=2018 ../
   cmake --build . --config Release

It would help you to understand.
I will try to build again with ext 1 and tell you how it goes.

Thank you for your pay attention.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/chadmv/cvwrap/issues/8#issuecomment-309326903>, or mute
the thread

**Comment by redbean** [comment]
oops. sry I just copy and pasted the command from your repo. 
I specified version as 2016
I will try the command as below
cmake -G "Visual Studio 14 2015 Win64" -DMAYA_VERSION=2016.5 ../

**Comment by redbean** [comment]
Ah. I have done to compile it. Successfully Made mll!!

My problem was as below

My devkit was for ext2
which I downloaded it from https://apps.autodesk.com/MAYA/ko/Detail/Index?id=5595681162321676385&os=Win64&appLang=en

I switched my devkit to 2016.1 from 
https://github.com/autodesk-adn/Maya-devkit

Thank you for your help Chad.


PS: I have a warning 
  E:\cvwrap\src\common.cpp(215): warning C4838: conversion from 'unsigned int' to 'int' requires a n                                                                           arrowing conversion [E:\cvwrap\build2016\src\cvwrap.vcxproj]

Is it ok to ignore this warning??
